File: frontend/src/utils/slicerProfileResolve.ts
// Printer-aware @BBL profile resolution for Spoolman-stored slicer presets.
// Spoolman keeps the @BBL marker but omits the printer-model token; append it
// back when the active printer is known. Mirror of
// backend/app/utils/slicer_profile_resolve.py (shared test vectors).

const BBL_MARKER = '@BBL';
const NOZZLE_SUFFIX_RE = /\s+\d+(?:\.\d+)?\s*nozzle\s*$/i;
const SETTING_ID_RE = /^[A-Za-z0-9]+$/;

// PRINTER_MODEL_MAP uses display names (e.g. "A1 Mini"); Bambu cloud process
// presets suffix @BBL with compact tokens (e.g. "A1M").
const BBL_PRESET_TAG_OVERRIDES: Record<string, string> = {
  'A1 Mini': 'A1M',
};

function bblPresetTag(token: string): string {
  return BBL_PRESET_TAG_OVERRIDES[token] ?? token;
}

/** Alternate @BBL suffixes for the same printer (cloud A1M vs bundled A1 Mini). */
const BBL_PRESET_TAG_ALTERNATES: Record<string, string> = {
  A1M: 'A1 Mini',
  'A1 Mini': 'A1M',
};

function bblTagAfterMarker(name: string): string | null {
  const idx = bblMarkerIndex(name);
  if (idx < 0) return null;
  const tag = name.slice(idx + BBL_MARKER.length).trim();
  return tag || null;
}

function swapBblTag(name: string, newTag: string): string {
  const idx = bblMarkerIndex(name);
  if (idx < 0) return name;
  return `${name.slice(0, idx + BBL_MARKER.length).trimEnd()} ${newTag}`;
}

/** Primary resolved names plus equivalent @BBL suffix variants, in order. */
function expandBblTagCandidates(candidates: string[]): string[] {
  const out: string[] = [];
  const seen = new Set<string>();
  for (const candidate of candidates) {
    if (!seen.has(candidate)) {
      out.push(candidate);
      seen.add(candidate);
    }
    const tag = bblTagAfterMarker(candidate);
    const alt = tag ? BBL_PRESET_TAG_ALTERNATES[tag] : null;
    if (!alt) continue;
    const variant = swapBblTag(candidate, alt);
    if (!seen.has(variant)) {
      out.push(variant);
      seen.add(variant);
    }
  }
  return out;
}

export type PrinterModelMap = Record<string, string>;

function bblMarkerIndex(name: string): number {
  return name.toUpperCase().indexOf(BBL_MARKER);
}

function looksLikeSettingId(name: string): boolean {
  return SETTING_ID_RE.test(name) && bblMarkerIndex(name) < 0;
}

/** Split a stored profile value into its comma-separated candidates. */
export function splitProfileCandidates(raw: string | null | undefined): string[] {
  if (!raw) return [];
  return raw
    .split(',')
    .map((part) => part.trim())
    .filter((part) => part.length > 0);
}

/** True when `name` carries `@BBL` followed by a model token. */
export function hasBblPrinterTag(name: string): boolean {
  if (!name) return false;
  const idx = bblMarkerIndex(name);
  if (idx < 0) return false;
  return name.slice(idx + BBL_MARKER.length).trim().length > 0;
}

/** Reduce a fully-qualified preset name to its printer-agnostic base. */
export function stripBblPrinterTag(name: string): string {
  if (!name) return '';
  const idx = bblMarkerIndex(name);
  if (idx < 0) return name.trim();
  return name.slice(0, idx + BBL_MARKER.length).trimEnd();
}

/** Normalise a model / printer-preset string to its `@BBL` short token. */
function canonicalToken(printerModel: string | null | undefined, models?: PrinterModelMap): string | null {
  if (!printerModel) return null;
  let cleaned = printerModel.trim();
  if (cleaned.startsWith('# ')) cleaned = cleaned.slice(2).trim();
  cleaned = cleaned.replace(NOZZLE_SUFFIX_RE, '').trim();
  if (!cleaned) return null;
  if (models && cleaned in models) return bblPresetTag(models[cleaned]);
  const stripped = cleaned.replace(/^Bambu Lab\s+/i, '').trim();
  return stripped ? bblPresetTag(stripped) : null;
}

/** Append the active printer's `@BBL` token to a printer-agnostic base. */
export function appendBblPrinterTag(
  baseName: string,
  printerModel: string | null | undefined,
  models?: PrinterModelMap,
): string {
  if (!baseName) return baseName;
  if (looksLikeSettingId(baseName)) return baseName;
  if (hasBblPrinterTag(baseName)) return baseName;

  const token = canonicalToken(printerModel, models);
  if (!token) return baseName;

  const idx = bblMarkerIndex(baseName);
  if (idx < 0) return `${baseName.trimEnd()} ${BBL_MARKER} ${token}`;
  const head = baseName.slice(0, idx + BBL_MARKER.length).trimEnd();
  return `${head} ${token}`;
}

/** Resolve a stored profile value into printer-qualified candidates, in order. */
export function resolveProfileForPrinter(
  raw: string | null | undefined,
  printerModel: string | null | undefined,
  models?: PrinterModelMap,
): string[] {
  return splitProfileCandidates(raw).map((candidate) => appendBblPrinterTag(candidate, printerModel, models));
}

function normalizeForMatch(name: string): string {
  return name
    .replace(/^#\s*/, '')
    .replace(/\s+/g, ' ')
    .trim()
    .toLowerCase();
}

/**
 * Find a unified preset for a spool's stored profile on the active printer.
 * Tries each resolved candidate by id / full name first, then by printer-agnostic base.
 */
export function matchPresetByProfile<T extends { id: string; name: string }>(
  presets: readonly T[],
  storedProfile: string | null | undefined,
  printerModel: string | null | undefined,
  models?: PrinterModelMap,
): T | null {
  const candidates = expandBblTagCandidates(
    resolveProfileForPrinter(storedProfile, printerModel, models),
  );
  if (candidates.length === 0) return null;

  // Prefer the printer-qualified name (or setting id) over a looser base match.
  for (const candidate of candidates) {
    const fullKey = normalizeForMatch(candidate);
    const isSettingId = looksLikeSettingId(candidate);
    for (const preset of presets) {
      if (isSettingId && preset.id.toLowerCase() === candidate.toLowerCase()) return preset;
      if (normalizeForMatch(preset.name) === fullKey) return preset;
    }
  }

  for (const candidate of candidates) {
    const baseKey = normalizeForMatch(stripBblPrinterTag(candidate));
    for (const preset of presets) {
      if (normalizeForMatch(stripBblPrinterTag(preset.name)) === baseKey) return preset;
    }
  }
  return null;
}

File: backend/app/utils/slicer_profile_resolve.py
"""@BBL printer-token strip/append for Spoolman-stored slicer profile names.

Spoolman keeps the ``@BBL`` marker but omits the printer-model token; append it
back when the active printer is known. Kept in sync with
``frontend/src/utils/slicerProfileResolve.ts`` (shared test vectors).
"""

from __future__ import annotations

import re

from backend.app.utils.printer_models import normalize_printer_model

_BBL_MARKER = "@BBL"
_NOZZLE_SUFFIX_RE = re.compile(r"\s+\d+(?:\.\d+)?\s*nozzle\s*$", re.IGNORECASE)
_SETTING_ID_RE = re.compile(r"^[A-Za-z0-9]+$")
# PRINTER_MODEL_MAP uses "A1 Mini"; Bambu cloud process presets use @BBL A1M.
_BBL_PRESET_TAG_OVERRIDES: dict[str, str] = {"A1 Mini": "A1M"}


def split_profile_candidates(raw: str | None) -> list[str]:
    """Split a stored profile value into comma-separated candidates, trimmed."""
    if not raw:
        return []
    return [part.strip() for part in raw.split(",") if part.strip()]


def _bbl_marker_index(name: str) -> int:
    """Index of the ``@BBL`` marker (case-insensitive), or -1 when absent."""
    return name.upper().find(_BBL_MARKER)


def _looks_like_setting_id(name: str) -> bool:
    """True for bare ids like ``GFSL05`` / ``PFUSabc`` (no spaces, no @)."""
    return bool(_SETTING_ID_RE.match(name)) and _bbl_marker_index(name) < 0


def has_bbl_printer_tag(name: str) -> bool:
    """True when ``@BBL`` is followed by a printer model token (not a bare base)."""
    if not name:
        return False
    idx = _bbl_marker_index(name)
    if idx < 0:
        return False
    rest = name[idx + len(_BBL_MARKER) :].strip()
    return bool(rest)


def strip_bbl_printer_tag(name: str) -> str:
    """Remove the printer-model token after ``@BBL``; keep the marker on the base."""
    if not name:
        return ""
    idx = _bbl_marker_index(name)
    if idx < 0:
        return name.strip()
    head = name[: idx + len(_BBL_MARKER)].rstrip()
    return head


def append_bbl_printer_tag(base_name: str, printer_model: str | None) -> str:
    """Append the active printer's model token after ``@BBL`` on a base name.

    Bare setting ids (``GFSL05``) and names that already carry a printer token
    are returned unchanged. When ``printer_model`` cannot be resolved, returns
    ``base_name`` unchanged.
    """
    if not base_name:
        return base_name
    if _looks_like_setting_id(base_name):
        return base_name
    if has_bbl_printer_tag(base_name):
        return base_name

    token = _canonical_token(printer_model)
    if not token:
        return base_name

    idx = _bbl_marker_index(base_name)
    if idx < 0:
        return f"{base_name.rstrip()} {_BBL_MARKER} {token}"
    head = base_name[: idx + len(_BBL_MARKER)].rstrip()
    return f"{head} {token}"


def resolve_profile_for_printer(raw: str | None, printer_model: str | None) -> list[str]:
    """Apply ``append_bbl_printer_tag`` to each comma-separated candidate, in order."""
    return [append_bbl_printer_tag(candidate, printer_model) for candidate in split_profile_candidates(raw)]


def _bbl_preset_tag(token: str) -> str:
    return _BBL_PRESET_TAG_OVERRIDES.get(token, token)


def _canonical_token(printer_model: str | None) -> str | None:
    """Normalise a printer model or preset name to a canonical ``@BBL`` short token."""
    if not printer_model:
        return None
    cleaned = printer_model.strip()
    if cleaned.startswith("# "):
        cleaned = cleaned[2:].strip()
    cleaned = _NOZZLE_SUFFIX_RE.sub("", cleaned).strip()
    if not cleaned:
        return None
    model = normalize_printer_model(cleaned)
    return _bbl_preset_tag(model) if model else None
